attestation-service/attestation-service/src/policy_engine/opa/mod.rs

```diff
@@ -63,51 +63,76 @@ impl PolicyEngine for OPA {
+                let mut hasher = sha2::Sha384::new();
+                hasher.update(&policy);
+                let hex = hasher.finalize().to_vec();
+                hex::encode(hex)
+            };
+
+            let policy_go = GoString {
+                p: policy.as_ptr() as *const c_char,
+                n: policy.len() as isize,
+            };
+
+            let reference = serde_json::json!({ "reference": reference_data_map }).to_string();
+
+            let reference_go = GoString {
+                p: reference.as_ptr() as *const c_char,
+                n: reference.len() as isize,
+            };
+
+            let input_go = GoString {
+                p: input.as_ptr() as *const c_char,
+                n: input.len() as isize,
+            };
+
+            // Call the function exported by cgo and process the returned decision
+            let decision_buf: *mut c_char =
+                unsafe { evaluateGo(policy_go, reference_go, input_go) };
+            let decision_str: &CStr = unsafe { CStr::from_ptr(decision_buf) };
+            let policy_res = decision_str.to_str()?.to_string();
+            debug!("Evaluated: {}", policy_res);
+            if policy_res.starts_with("Error::") {
+                bail!("OPA verification failed: {policy_res}");
+            }
 
-        let input_go = GoString {
-            p: input.as_ptr() as *const c_char,
-            n: input.len() as isize,
-        };
+            // If a clear approval opinion is given in the evaluation report,
+            // the rejection information will be reflected in the evaluation failure return value.
+            let res_kv: Value = serde_json::from_str(&policy_res)?;
 
-        // Call the function exported by cgo and process the returned decision
-        let decision_buf: *mut c_char = unsafe { evaluateGo(policy_go, reference_go, input_go) };
-        let decision_str: &CStr = unsafe { CStr::from_ptr(decision_buf) };
-        let res = decision_str.to_str()?.to_string();
-        debug!("Evaluated: {}", res);
-        if res.starts_with("Error::") {
-            return Err(anyhow!(res));
-        }
+            // only if there is a field named `allow` in the evaluation report and
+            // it is false, the evaluation fails. Otherwise the evaluation will be
+            // treated as succees.
+            let allow = res_kv
+                .get("allow")
+                .map(|a| a.as_bool())
+                .map(|a| a.unwrap_or(true))
+                .unwrap_or(true);
 
-        // If a clear approval opinion is given in the evaluation report,
-        // the rejection information will be reflected in the evaluation failure return value.
-        let res_kv: Value = serde_json::from_str(&res)?;
-        if let Some(allow) = res_kv["allow"].as_bool() {
             if !allow {
-                bail!("Untrusted TEE evidence")
+                bail!("TEE evidence does not pass policy {policy_id}, reason: {policy_res}");
```

Comment:
I'm not sure this is the best way to handle this. It might be better to return a token that says that a policy has not passed evaluation. It's a bit hard to know. Maybe that would increase the complexity of the KBS.

If we stick with this approach we should make sure to document it well for clients who are trying to write policies.

Reply:
In the initial implementation I embed the policy validation results into the result token. Suggested by @jepio  https://github.com/confidential-containers/kbs/pull/240#discussion_r1411795084   I think it is right that if any policy fails to check, the verifiecation should fail.

Reply:
So there are two options for now
1. Provision token only if all policies pass. less flexible for policy but easy to use token.
2. Provision token regardless of any policy enforcement result. More flexible for policy but needs extra logic to use token.

right?

Now I am ok with both.

wdyt? @fitzthum @jepio

Reply:
Yeah I think either one is fine. The first option sort of implies that the AS policies are going to return binary values, which isn't necessarily the case. The second option will likely increase the complexity of the KBS policies. I guess we can pick one, merge it, and see how we feel about it down the road.